### ISSUE TITLE
Miscompilation of functions w/ defer statements

### ISSUE THREAD

**Issue by bpowers** [gopherjs/gopherjs#493]
Hello,

I have a [fork of gopherjs](https://github.com/bpowers/browsix-gopherjs) with system call support in the browser, where `syscall.Syscall` waits on a blocking channel read.  We're seeing a problem with functions that have a defer statement that blocks (like `os.File.Close()`), and a call of a function that returns no parameters immediately before a return statement, like the following:

```Go
package main

import (
	"log"
	"os"
)

func nullLog(fmt string, args ...interface{}) {

}

func Load(name string) (err error) {

	err = nil

	var file *os.File
	if file, err = os.Open("/"); err != nil {
		return
	}

	defer file.Close()

	// the problem happens when: a void function is invoked
	// immediately before a return statement within a function
	// that has a deferred function that blocks.
	nullLog("Load completed.")
	return err
}

func main() {

	err := Load("example")
	if err != nil {
		log.Fatalf("uh oh")
	}

	log.Printf("loaded example")
}
```

(this runs OK under node with the node-syscall package, I think because that does regular sync syscalls on the main node thread).

The issue is that when the defer statement resumes, in the generated JS a `case` statement + increment of `$s` has not been emitted between the function call to `nullLog` (in this case) and `return err` -- so execution of `Load()` jumps to immedately before the call to `nullLog` instead of after, that function call is reissued _even though it didn't block itself_, and bad things happen.

Please let me know if you need anymore info, I've misdiagnosed anything, or have any questions!  Thanks!

CC'ing @b1101 who initially saw this.

**Comment by shurcooL** [comment]
This could be related to #381.

**Comment by bpowers** [comment]
Yes - absolutely related, maybe even a duplicate.

**Comment by bpowers** [comment]
Somehow didn't see that in my glance of the open isues

**Comment by shurcooL** [comment]
It's not easy to spot, I just happen to be closely familiar with it because IMO this is the highest priority known issue in GopherJS right now, aside from Go 1.7 support. I hope @neelance can get a chance to look at it sometime.

It has already come up in various forms many times - here, in #381, and likely #426 as well.

**Comment by bpowers** [comment]
@shurcooL I think that the line on https://github.com/gopherjs/gopherjs/blob/master/compiler/expressions.go#L776
needs to change from
```Go
if c.Blocking[e] {
```

to something like
```Go
if c.Blocking[e] || c.hasBlockingDeferStmt() {
```
